walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,126 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if ((int) request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    if (events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<Event> updatedEvents = addUpdatedEvents(events, request);
+    // If ALL the events are for non-meeting attendees, just return list of whole day.
+    if (updatedEvents.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<TimeRange> slotsAvailable = findSlotsAvailable(updatedEvents, request);
+
+    return slotsAvailable;
+  }
+
+  private List<Event> addUpdatedEvents(Collection<Event> events, MeetingRequest request) {
+    List<Event> updatedEvents = new ArrayList<>();
+    List<Event> eventsList = new ArrayList<>(events);
+
+    Collections.sort(eventsList, new Comparator<Event>() {
+      @Override
+      public int compare(Event a, Event b) {
+        return Long.compare(a.getWhen().start(), b.getWhen().start());
+      }
+    });
+
+    // Add earliest VALID event to updatedEvents first, to use as a comparison for subsequent event.
+    int firstEventNum = 0;
+    boolean findingFirstEvent = true;
+    Collection<String> meetingAttendees = request.getAttendees();
+    while (findingFirstEvent && firstEventNum < events.size()) {
+      Event event = eventsList.get(firstEventNum);
+      Collection<String> eventAttendees = event.getAttendees();
+
+      // If every event attendees is NOT a meeting attendee, ignore the event.
+      if (!meetingAttendees.containsAll(eventAttendees)) {
+        firstEventNum++;
+        continue;
+      }
+
+      firstEventNum++;
+      updatedEvents.add(event);
+      findingFirstEvent = false;
+    }
```

Comment:
Line 56 - 73 can be abstracted into a helper method to make the code more readable -- the method takes in `updatedEvents` (modifies it inside the method) and returns `firstEventNum` as a result.

Also note that `firstEventNum` might not be the best name, consider using `index` or `count` instead of `num` to better represent the value.